broker/cloud_functions/filter_purity/main.py

```diff
@@ -0,0 +1,124 @@
+
+schema_map = schema_maps.load_schema_map(SURVEY, TESTID)
+
+
+def is_pure(alert, schema_map):
+    """Adapted from: https://zwickytransientfacility.github.io/ztf-avro-alert/filtering.html
+
+    Quoted from the source:
+
+    ZTF alert streams contain an nearly entirely unfiltered stream of all
+    5-sigma (only the most obvious artefacts are rejected). Depending on your
+    science case, you may wish to improve the purity of your sample by filtering
+    the data on the included attributes.
+
+    Based on tests done at IPAC (F. Masci, priv. comm), the following filter
+    delivers a relatively pure sample.
+    """
+    source = alert[schema_map['source']]
+
+    rb = (source['rb'] >= 0.65)  # RealBogus score
+
+    if schema_map['SURVEY'] == 'decat':
+        is_pure = rb
+
+    elif schema_map['SURVEY'] == 'ztf':
+        nbad = (source['nbad'] == 0)  # num bad pixels
+        fwhm = (source['fwhm'] <= 5)  # Full Width Half Max, SExtractor [pixels]
+        elong = (source['elong'] <= 1.2)  # major / minor axis, SExtractor
+        magdiff = (abs(source['magdiff']) <= 0.1)  # aperture - psf [mag]
+        is_pure = (rb and nbad and fwhm and elong and magdiff)
+
+    return is_pure
+
+
+
+def run(msg: dict, context) -> None:
+    """Filter alerts for purity, publish results.
+
+    For args descriptions, see:
+    https://cloud.google.com/functions/docs/writing/background#function_parameters
+
+    Args:
+        msg: Pub/Sub message data and attributes.
+            `data` field contains the message data in a base64-encoded string.
+            `attributes` field contains the message's custom attributes in a dict.
+
+        context: The Cloud Function's event metadata.
+            It has the following attributes:
+                `event_id`: the Pub/Sub message ID.
+                `timestamp`: the Pub/Sub message publish time.
+                `event_type`: for example: "google.pubsub.topic.publish".
+                `resource`: the resource that emitted the event.
+    """
+    # logger.log_text(f"{type(msg['data'])}', severity='DEBUG")
+    # logger.log_text(f"{msg['data']}', severity='DEBUG")
+
+    alert_dict = data_utils.decode_alert(
+        base64.b64decode(msg["data"]), drop_cutouts=True, schema_map=schema_map
+    ) # this decodes the alert
+    attrs = {
+        schema_map["objectId"]: str(alert_dict[schema_map["objectId"]]),
+        schema_map["sourceId"]: str(alert_dict[schema_map["sourceId"]]),
+    } # this gets the custom attr for filtering
+
+    # # run the alert through the filter.
+    # alert_is_pure = <set to boolean. True if alert passes purity filter, else False>
+    #
+    alert_is_pure = is_pure(alert_dict, schema_map) ## Should I use attrs instead of schema map?
+
+    # # if it passes the filter, publish to Pub/Sub:
+    # gcp_utils.publish_pubsub(ps_topic, alert_dict, attrs=attrs)
+    #
+    if alert_is_pure:
+        gcp_utils.publish_pubsub(ps_topic, alert_dict, attrs=attrs)
```

Comment:
We want to publish the results of `is_pure` along with the original alert. Follow this format, but use is_pure instead of SuperNNova:
```python
gcp_utils.publish_pubsub(
            ps_topic, {"alert": alert_dict, "SuperNNova": snn_dict}, attrs=attrs
        )
```